src/QMCWaveFunctions/detail/CUDA/cuBLAS_LU.cu

```diff
@@ -0,0 +1,253 @@
+  log_value.y = (lud < 0) * M_PI;
+  return log_value;
+}
+
+__device__ cuDoubleComplex complexDetLog(const cuDoubleComplex lu_diag, const int n_index, const int pivot)
+{
+  cuDoubleComplex diag;
+  double pivot_factor = 1 - 2 * (pivot != n_index + 1);
+  diag.x              = lu_diag.x * pivot_factor;
+  diag.y              = lu_diag.y * pivot_factor;
+  cuDoubleComplex log_value;
+  log_value.x = log(sqrt(diag.x * diag.x + diag.y * diag.y));
+  log_value.y = atan2(diag.y, diag.x);
+  return log_value;
+}
+
+/** computes the log determinant using the output of LU factorization.
+ *
+ *  logdets are assumed to be zeroed on launch of kernel.
+ */
+template<int COLBS, typename T>
+__global__ void computeLogDet_kernel(const int n,
+                                     const int lda,
+                                     T** mat_lus,
+                                     const int* const pivots,
+                                     cuDoubleComplex* logdets)
+{
+  const int iw         = blockIdx.x;
+  const int block_num  = blockIdx.y;
+  const T* lu_iw       = mat_lus[iw];
+  const int* pivots_iw = pivots + iw * n;
+  __shared__ cuDoubleComplex logdet_vals[COLBS];
+  logdet_vals[threadIdx.x].x = 0.0;
+  logdet_vals[threadIdx.x].y = 0.0;
+  T lu_diag;
+  int n_index = threadIdx.x + block_num * COLBS;
+  if (n_index < n)
+  {
+    lu_diag                  = *(lu_iw + n_index * lda + n_index);
+    logdet_vals[threadIdx.x] = complexDetLog(lu_diag, n_index, *(pivots_iw + n_index));
+  }
+  // insure that when we reduce logdet_vals all the threads in the block are done.
+  __syncthreads();
+  if (threadIdx.x == 0)
+  {
+    cuDoubleComplex block_sum_log_det;
+    block_sum_log_det.x = 0.0;
+    block_sum_log_det.y = 0.0;
+    for (int iv = 0; iv < COLBS; ++iv)
+    {
+      block_sum_log_det.x += logdet_vals[iv].x;
+      block_sum_log_det.y += logdet_vals[iv].y;
+    }
+    // No atomicAdd in cuda 10 for cuComplex/cuDoubleComplex
+    atomicAdd((double*)&(logdets[iw].x), block_sum_log_det.x);
+    atomicAdd((double*)&(logdets[iw].y), block_sum_log_det.y);
+  }
+}
+
+/** Calculates logdets using LU_diags and pivots
+ *  \param[in] LU_mat - the LU output from cublasXgetrfBatched
+ *  \param[out] LU_diags - the LU_diags from the LU
+ *  \param[in] batch_size - no a big deal here.
+ */
+template<typename TMAT, typename T>
+cudaError_t computeLogDet_batched_impl(cudaStream_t& hstream,
+                                       const int n,
+                                       const int lda,
+                                       TMAT** LU_mat,
+                                       const int* pivots,
+                                       T* logdets,
+                                       const int batch_size)
+{
+  // Perhaps this should throw an exception. I can think of no good reason it should ever happen other than
+  // developer error.
+  if (batch_size == 0 || n == 0)
+    return cudaSuccess;
+
+  CUDAfill_n(logdets, batch_size, {0.0, 0.0});
```

Comment:
I'm more convinced to do one full matrix per block. CUDAfill_n was never intended to be fast.

Reply:
I'm sure I can write a `__device__` function that will be fast but I think that needs to be in another PR.